clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -947,6 +947,8 @@ private void throwIfProducerClosed() {
             // for other exceptions throw directly
         } catch (ApiException e) {
             log.debug("Exception occurred during message send:", e);
+            if (e instanceof BufferExhaustedException)
```

Comment:
It's fine to keep compatibility for now. However, I do wonder if we should just throw ApiException to the caller since the producer can block for max.block.ms. This needs a KIP discussion.

Reply:
I haven't looked at it, but do all ApiExceptions that may be caught here observer max.block.ms?

Reply:
I was thinking that all retriable errors should observe max.block.ms if not already. For non-retirable errors, just throw the exception immediately.

Reply:
I'll do some digging on this topic. But if I am not misunderstanding you, we can handle that in a separate change and merge this for now, right?